operator/pkg/ciliumidentity/cache.go

```
@@ -0,0 +1,331 @@
+			break
+		}
+	}
+}
+
+func (c *CIDState) LookupByID(id string) (*key.GlobalIdentity, bool) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	k, exists := c.idToLabels[id]
+	return k, exists
+}
+
+func (c *CIDState) LookupByKey(k *key.GlobalIdentity) (string, bool) {
+	if k == nil {
+		return "", false
+	}
+
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	secIDs, exists := c.labelsToID[k.GetKey()]
+	if !exists {
+		return "", false
+	}
+
+	var id string
+	if secIDs != nil {
+		id = secIDs.selectedID
+	}
+	return id, true
+}
+
+type CIDUsageInPods struct {
+	podToCID      map[string]string
+	cidUsageCount map[string]int
+
+	mu lock.RWMutex
+}
+
+func NewCIDUsageInPods() *CIDUsageInPods {
+	return &CIDUsageInPods{
+		podToCID:      make(map[string]string),
+		cidUsageCount: make(map[string]int),
+	}
+}
+
+// AssignCIDToPod updates the pod to CID map and increments the CID usage.
+// It also decrements the previous CID usage and returns the CID name of
+// previously set CID and its usage count after decrementing the CID usage.
+// The return values are used to track when old CIDs are no longer used.
+func (c *CIDUsageInPods) AssignCIDToPod(podName, cidName string) (string, int) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	var prevCIDUsageCount int
+	prevCIDName, exists := c.podToCID[podName]
+	if exists {
+		if cidName == prevCIDName {
+			return cidName, c.cidUsageCount[cidName]
+		}
+
+		prevCIDUsageCount = c.decrementUsage(prevCIDName)
+	}
+
+	c.podToCID[podName] = cidName
+	c.cidUsageCount[cidName]++
+
+	return prevCIDName, prevCIDUsageCount
+}
+
+// RemovePod removes the pod from the pod to CID map, decrements the CID usage
+// and returns the CID name and its usage count after decrementing the usage.
+// The return values are used to track when old CIDs are no longer used.
+func (c *CIDUsageInPods) RemovePod(podName string) (string, int) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	cidName, exists := c.podToCID[podName]
+	if !exists {
+		return "", 0
```

In case `podName` does not exist in `c.podToCID` I think it should be better to return an explicit error.